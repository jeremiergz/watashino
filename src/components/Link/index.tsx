import Anchor, { AnchorProps } from 'components/Anchor';
import withoutStylingProps from 'components/hocs/withoutStylingProps';
import { useTheming } from 'components/providers/ThemeProvider';
import { Link as BaseGatsbyLink } from 'gatsby';
import { OutboundLink as BaseOutboundLink } from 'gatsby-plugin-google-analytics';
import React from 'react';
import { Routes } from 'utils/Routes';

const GatsbyLink = withoutStylingProps(BaseGatsbyLink);
const OutboundLink = withoutStylingProps(BaseOutboundLink);

const Link: React.FC<LinkProps> = ({ children, external, partiallyActive = false, to, ...rest }) => {
  const { theme } = useTheming();
  const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
  const target = isTouchDevice ? '_self' : '_blank';
  const linkProps = external
    ? {
        as: OutboundLink,
        href: to,
        rel: 'noopener noreferrer',
        target,
      }
    : {
        activeStyle: partiallyActive ? { color: theme.colors.primary } : {},
        as: GatsbyLink,
        partiallyActive: partiallyActive && (to !== Routes.index ? true : false),
        to,
      };
  return (
    <Anchor {...linkProps} {...rest}>
      {children}
    </Anchor>
  );
};

Link.displayName = 'Link';

export type LinkProps = AnchorProps & {
  external?: boolean;
  partiallyActive?: boolean;
  to: string;
};
export default Link;
