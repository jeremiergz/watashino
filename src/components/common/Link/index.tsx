import { Link as BaseGatsbyLink } from 'gatsby';
import React from 'react';
import theme from '../../../theme';
import withoutStylingProps from '../../hocs/withoutStylingProps';
import Anchor, { AnchorProps } from '../Anchor';

type LinkProps = AnchorProps & {
    external?: boolean;
    partiallyActive?: boolean;
    to: string;
};

const GatsbyLink = withoutStylingProps(BaseGatsbyLink);

const Link = ({ children, external, partiallyActive = false, to, ...rest }: LinkProps) => {
    const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
    const target = isTouchDevice ? '_self' : '_blank';
    const linkProps = external
        ? {
              href: to,
              rel: 'noopener noreferrer',
              target,
          }
        : {
              activeStyle: { color: theme.colors.primary },
              as: GatsbyLink,
              partiallyActive: partiallyActive && (to !== '/' ? true : false),
              to,
          };
    return (
        <Anchor {...linkProps} {...rest}>
            {children}
        </Anchor>
    );
};

Link.displayName = 'Link';

export { LinkProps };
export default Link;
