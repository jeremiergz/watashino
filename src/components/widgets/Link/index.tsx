import { Link as BaseGatsbyLink } from 'gatsby';
import React from 'react';
import styled from 'styled-components';
import { variant } from 'styled-system';
import theme from '../../../theme';
import withoutStylingProps from '../../hocs/withoutStylingProps';
import BaseBox, { BoxProps } from '../../primitives/Box';

type LinkProps = BoxProps & {
    external?: boolean;
    to: string;
    variant?: 'underlined';
};

const Box = styled(BaseBox)`
    text-decoration: none;
    transition: color 100ms ease-in-out;
    .gatsby-image-wrapper {
        transition: filter 100ms ease-in-out;
    }
    :hover {
        color: ${theme.colors.primary};
        .gatsby-image-wrapper {
            filter: brightness(0.75);
        }
    }
    ${variant({
        variants: {
            underlined: {
                textDecoration: 'underline',
            },
        },
    })}
`;

const GatsbyLink = withoutStylingProps(BaseGatsbyLink);

const Link = ({ children, external, to, ...rest }: LinkProps) => {
    const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
    const target = isTouchDevice ? '_self' : '_blank';
    const linkProps = external
        ? {
              href: to,
              rel: 'noopener noreferrer',
              target,
          }
        : {
              activeStyle: { color: theme.colors.primary },
              partiallyActive: to !== '/' ? true : false,
              to,
          };
    return (
        <Box as={external ? 'a' : GatsbyLink} color="text" fontWeight="semi-bold" {...linkProps} {...rest}>
            {children}
        </Box>
    );
};

Link.displayName = 'Link';

export { LinkProps };
export default Link;
