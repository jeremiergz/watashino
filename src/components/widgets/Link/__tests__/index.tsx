import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Link from '..';

describe('components/widgets/Link Test Suite', () => {
    it('matches snapshot', () => {
        const jsx = (
            <Link to="/test-page" variant="underlined">
                Link Component
            </Link>
        );
        expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
    });
});
