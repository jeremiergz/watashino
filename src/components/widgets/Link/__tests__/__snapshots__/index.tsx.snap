// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/widgets/Link Test Suite matches snapshot 1`] = `
<a
  className="sc-bdVaJa sc-bwzfXH fVKJCI"
  href="/test-page"
  variant="underlined"
>
  Link Component
</a>
`;
